src/hypertable_cache.c

```diff
@@ -129,24 +130,51 @@ ts_hypertable_cache_invalidate_callback(void)
 
 /* Get hypertable cache entry. If the entry is not in the cache, add it. */
 TSDLLEXPORT Hypertable *
-ts_hypertable_cache_get_entry(Cache *cache, Oid relid)
+ts_hypertable_cache_get_entry(Cache *const cache, const Oid relid, const bool missing_ok)
 {
+	Hypertable *ht;
+
 	if (!OidIsValid(relid))
-		return NULL;
+		ht = NULL;
+	else
+		ht = ts_hypertable_cache_get_entry_with_table(cache, relid, NULL, NULL);
+
+	if (!missing_ok && ht == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_HYPERTABLE_NOT_EXIST),
+				 errmsg("table \"%s\" is not a hypertable", get_rel_name(relid))));
+
+	return ht;
+}
+
+TSDLLEXPORT Hypertable *
+ts_hypertable_cache_get_entry_missing_ok(Cache *const cache, const Oid relid)
+{
+	return ts_hypertable_cache_get_entry(cache, relid, true);
+}
 
-	return ts_hypertable_cache_get_entry_with_table(cache, relid, NULL, NULL);
+/*
+ * Returns cache into the argument and hypertable as the function result.
+ * If hypertable is not found, fails with an error.
+ */
+Hypertable *
+ts_hypertable_cache_get_entry_or_error(const Oid relid, Cache **const cache)
+{
+	*cache = ts_hypertable_cache_pin();
```

Comment:
Always forcing cache pinning here makes this function useless across multiple calls using the same cache.

I think you should either 

* skip this cache pinning functionality, or
* check if the cache pointer is already set and not pin again (although this is odd, because it forces always passing in a valid cache in that case), or
* provide a complementary function that uses an existing cache so that this function can be called multiple times for the same cache.

Reply:
> * provide a complementary function that uses an existing cache so that this function can be called multiple times for the same cache.

The function exists as in used. It's called `ts_hypertable_cache_get_entry`.

`ts_hypertable_cache_get_entry_or_error` is only used in the cases where `ts_hypertable_cache_pin` was present. My changes don't change any semantics and if you substitute the function with the body, you will get the original code.